Add $quantityWithUnitFormat parameter to Quantity[Html]Formatter

thiemowmde opened the pull request:
The QuantityFormatter constructor does have a breaking change anyway. I'm reordering the parameters so that options are first, then the two formatters for the number and the unit, then the format string to combine the two.

[Bug: T111186](http://phabricator.wikimedia.org/T111186)

brightbyte commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -63,14 +63,25 @@ class QuantityFormatter extends ValueFormatterBase {
-	 * @param FormatterOptions|null $options
+	 * @param string|null $quantityWithUnitFormat Warning, this must be under the control of the
+	 * application, not under the control of the user, because it allows HTML injections in
+	 * subclasses that return HTML.
```
Please describe the expected format of this string

brightbyte commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -80,12 +91,22 @@ public function __construct(
+	 *
+	 * @return string
+	 */
+	protected function getQuantityWithUnitFormat() {
```
Why do we need this method? Why is it protected? What can be achieved by overriding it?

brightbyte commented:
Seems sensible, though not really needed right now.

brightbyte commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -63,14 +63,25 @@ class QuantityFormatter extends ValueFormatterBase {
+	 * subclasses that return HTML.
 	 */
 	public function __construct(
+		FormatterOptions $options = null,
```
Changing the order here makes this a breaking change. 
Without this, we could push this until after the major release.

thiemowmde replied:
No. It already is a breaking change.

brightbyte replied:
How? Did I miss something?

thiemowmde replied:
You merged it: https://github.com/DataValues/Number/pull/41/files#diff-8c77e0fb91ce0037fa8e2fd7b88a35ab

thiemowmde commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -63,14 +63,26 @@ class QuantityFormatter extends ValueFormatterBase {
+	 * HTML.
 	 */
 	public function __construct(
+		FormatterOptions $options = null,
```
I have split this change off to #45. Please merge #45 first and let me know, I will rebase #44 then.

brightbyte commented:
+2 from me. I would merge this (no need for #45 really), but travis sais no :(

thiemowmde commented:
Where does Travis say no?

thiemowmde commented:
Can you please merge #45?

thiemowmde commented:
Rebased and reworked QuantityHtmlFormatter a bit to not duplicate the "quantityWithUnitFormat" code in QuantityFormatter::formatQuantityValue, instead it overrides formatNumber and formatUnit now.
